napi_js_data.md

```diff
@@ -0,0 +1,758 @@
+
+#### Parameters
+- `[in]  env`: The environment that the API is invoked under
+- `[in]  type`: Scalar datatype of the elements within the TypedArray
+- `[in]  length`: Number of elements in the TypedArray
+- `[in]  arraybuffer`: ArrayBuffer underlying the typed array
+- `[in]  byte_offset`: The byte offset within the ArrayBuffer from which to start projecting the TypedArray
+- `[out] result`: A `napi_value` representing a JavaScript TypedArray
+
+#### Return value
+- `napi_ok` if the API succeeded.
+
+#### Description
+This API creates a JavaScript TypedArray object over an existing ArrayBuffer.
+TypedArray objects provide an array-like view over an underlying data buffer 
+where each element has the same underlying binary scalar datatype.
+
+It's required that (length * size_of_element) + byte_offset should be <= ArrayBuffer.ArrayBufferByteLength. 
+If not, a RangeError exception is raised.
+
+JavaScript TypedArray Objects are described in [Section 22.2](https://tc39.github.io/ecma262/#sec-typedarray-objects) 
+of the ECMAScript Language Specification.
+
+### *napi_get_array_length*
+
+#### Signature
+```
+napi_status napi_get_array_length(napi_env env, napi_value value, uint32_t* result)
+```
+
+#### Parameters
+- `[in]  env`: The environment that the API is invoked under
+- `[in]  value`: napi_value representing the JavaScript Array whose length is being queried
+- `[out] result`: uint32 representing length of the array
+
+#### Return value
+- `napi_ok` if the API succeeded.
+
+#### Description
+This API returns the length of an array.
+
+Array length is described in [Section 22.1.4.1](https://tc39.github.io/ecma262/#sec-properties-of-array-instances-length) 
+of the ECMAScript Language Specification.
+
+### *napi_get_arraybuffer_info*
+
+#### Signature
+```
+napi_status napi_get_arraybuffer_info(napi_env env,
+                                      napi_value arraybuffer,
+                                      void** data,
+                                      size_t* byte_length)
+```
+
+#### Parameters
+- `[in]  env`: The environment that the API is invoked under
+- `[in]  arraybuffer`: napi_value representing the ArrayBuffer being queried
+- `[out] data`: The underlying data buffer of the ArrayBuffer
+- `[out] byte_length`: Length in bytes of the underlying data buffer
+
+#### Return value
+- `napi_ok` if the API succeeded.
+
+#### Description
+This API is used to retrieve the underlying data buffer of an ArrayBuffer and it's length.
+WARNING: This API is dangerous. The lifetime of the underlying data buffer is managed by the 
+ArrayBuffer even after it's returned to you. The one possible safe to use this API is in conjunction with napi_create_reference, which can be used to guarantee that you control the lifetime of the ArrayBuffer. It's also safe to use the returned data buffer within the 
+same callback as long as you don't call any other APIs that might trigger a GC.
+
+<TODO: What happens to this buffer in case of a compacting GC?>
```

Comment:
More the latter I think. It's one of those cases where we might be taking an implicit dependency on current VM behavior- nothing really precludes the VM from moving the underlying buffer as part of a GC, in which case the user's access to the buffer is invalid. Using a handle might help future-proof this code at the cost of perf. Just something I thought of while writing the docs, not something that needs to be urgently addressed